### PR TITLE
CLN: Make iterators lazier

### PR DESCRIPTION
**Pull request by mroeschke**

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @mroeschke

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -706,7 +706,7 @@ def groups(self) -> dict[Hashable, Index]:
             return self.groupings[0].groups
         result_index, ids = self.result_index_and_ids
         values = result_index._values
-        categories = Categorical(ids, categories=np.arange(len(result_index)))
+        categories = Categorical(ids, categories=range(len(result_index)))
```
Comment (rhshadrach) [inline pandas/core/groupby/ops.py:709]: Don't we eventually need to make a NumPy array out of this? I haven't checked, but I imagine we're going to be passing this to `np.array`:

```
size = 100_000
%timeit np.arange(size)
23.5 µs ± 109 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit np.array(range(size))
4.77 ms ± 12.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Comment (WillAyd) [reply pandas/core/groupby/ops.py:709]: @mroeschke has been putting in a lot of work into deferring physical creation of indices and leveraging RangeIndex objects as much as possible, so this is in line with that larger body of work. 

If you measure through the lens of Index creation the benchmarks are different:

```python
In [16]: %timeit pd.Index(range(100_000))
2.45 µs ± 16.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [17]: %timeit pd.Index(np.arange(100_000))
30.8 µs ± 226 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Comment (rhshadrach) [reply pandas/core/groupby/ops.py:709]: Ah - we store as an Index. I should have checked, thanks.

Comment (mroeschke) [reply pandas/core/groupby/ops.py:709]: Thanks for the explanation @WillAyd. Sorry I probably should have commented about that in this PR

**Pull request merged**